### ISSUE TITLE
Likelihood vs KL balance

### ISSUE THREAD

**Issue by darleybarreto** [astirn/VariationalVariance#1]
Hi Mr. Stirn, great work on the code and the paper, really easy to follow and understand them. I have a question about the balance between the (log) Likelihood and KL terms, have you faced any problems with them? What's happening to me is the KL between the prior and the predicted Gamma is not decreasing, on the contrary, it is increasing (slowly), thus the predicted precision is very different from the prior, which messes the whole training. 
```python
ll = 0.5 * (expected_log_lambda - tf.math.log(2 * np.pi) - (y - mu) ** 2 * expected_lambda)
```
The `expected_lambda`, i.e., the mean of the Gamma, keeps decreasing, bringing `ll` down. The `expected_lambda` just keeps decreasing `ll`, and the KL is never minimized. Would you happen to have any insight on this?

PS: I am not actually running your code. I am working on orthogonal ideas, which enabled me to plug your work in, specifically the way the variance is reparameterized in terms of the Gamma for regression.

Thank you in advance.

**Comment by astirn** [comment]
Hey Darley,

Thanks for your interest. Sorry it has taken me a few days to reply. I am
overhauling a bunch of the code as we speak based on NeurIPS feedback. Can
I ask, which prior family are you considering (i.e. Standard, VAMP,
etc...)? Are you seeing LL decrease (or negative LL). If LL decreases, that
is obviously bad, and may warrant different learning rates? Remember the VI
objective is LL - DKL, so neither LL will be fully maximized nor DKL fully
minimized. Are you considering regression or VAE tasks? If the later
(although you could likely adapt it to the former), take a look at the Beta
VAE (Higgins), which affords you to scale the loss as  LL - beta * DKL such
that you can balance the objectives intuitively. Also, the standard prior
can act in the same manner but is more complicated to tune. Without knowing
more specifics, I would guess that if lambda and log lambda are decreasing,
in turn lowering the LL, perhaps your prior is too strong. Try using the
'mle' prior which ablates the KL divergence and see if you get
improvement--this would be an interesting diagnostic.

**Comment by darleybarreto** [comment]
Thank you for your reply!

> Can I ask, which prior family are you considering (i.e. Standard, VAMP,
etc...)?

It's equivalent to your Standard

> [...] and may warrant different learning rates?

I have tested with some values of lr, some values won't work (the big ones lie 0.1), and others (medium and small) will eventually give similar results, the difference it's their time to converge there.

>  Are you considering regression or VAE tasks? If the later (although you could likely adapt it to the former), take a look at the Beta VAE (Higgins), which affords you to scale the loss as  LL - beta * DKL such that you can balance the objectives intuitively.

It is regression, and I tweaked the beta parameter too :disappointed: 

> Also, the standard prior can act in the same manner but is more complicated to tune

I am using per-sample priors, so I am performing regression on the parameters of distributions, e.g., given labeled `mu` and `variance`, I use NNs to infer such parameters from the images.

> Try using the 'mle' prior which ablates the KL divergence and see if you get improvement

That's sound great, thank you so much for your time and advice!

**Comment by astirn** [comment]
Per-sample priors can be a bit tricky to conceptualize. Are you saying your
per-sample-precision generative process is:
For i in [N]:
   sample precision ~ p(precision; a_i, b_i)
   sample y | x ~ Normal(y | mu(x), precision)

As opposed to mine:
For i in [N]:
   sample precision ~ p(precision; a, b)
   sample y | x ~ Normal(y | mu(x), precision)

The critical difference is that a,b are global in mine. Let me know, I
might have some thoughts.

**Comment by darleybarreto** [comment]
I think I might have expressed myself incorrectly, for each `x \in X`, I have access to a prior (which I find in a pre-processing task). So it's basically 
```
for batch in samples:
   a = alpha(batch)
   b = beta(batch)
   sample precision ~ p(precision; a, b)
   sample y | batch ~ Normal(y | mu(batch), precision)
```
Where y is `mu` of my target `Normal(mu, global_precision)`

**Comment by darleybarreto** [comment]
I expressed myself in a wrong way, my apologies. My task is to predict per-sample parameters of a Gaussian (ideally `mu` and `precision`, but I am focusing on the last):
```
for samples in batches:
    mu = mu_net(samples)
    a = alpha_net(samples)
    b = beta_net(samples)

    precision ~ Gamma(precision; a, b)
    y | samples ~ Normal(y | mu, precision)
```

In my case I know how good or bad my `mu` and `precision` are, because I have computed per-sample priors in a pre-processing phase. Ideally `mu` would also come from a like `Normal( mu_net(samples), ones)`, but I am skipping it for now.
